circuit_passes/src/passes/loop_unroll/observer.rs

```diff
@@ -0,0 +1,290 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+use compiler::intermediate_representation::{ToSExp, UpdateId};
+use compiler::intermediate_representation::ir_interface::*;
+use crate::bucket_interpreter::env::Env;
+use crate::bucket_interpreter::error::{self, BadInterp};
+use crate::bucket_interpreter::memory::PassMemory;
+use crate::bucket_interpreter::observer::Observer;
+use crate::bucket_interpreter::{InterpreterFlags, LOOP_LIMIT};
+use crate::checked_insert;
+use crate::passes::{builders, GlobalPassData};
+use super::body_extractor::LoopBodyExtractor;
+use super::call_unroll_tree::Node;
+use super::loop_env_recorder::EnvRecorder;
+use super::map_like_trait::MapLike;
+use super::{
+    BlockBucketId, LoopBucketId, DEBUG_LOOP_UNROLL, EXTRACT_LOOP_BODY_TO_NEW_FUNC,
+    UNROLLED_BUCKET_LABEL,
+};
+
+#[derive(Debug, Default)]
+pub(super) struct LoopUnrollObserverResult {
+    /// Structure that owns all BlockBucket created by unrolling a loop. Keyed by the ID of
+    /// the BlockBucket so the other structures that organize and cache them can find them.
+    /// Note: uses Rc internally or else recursively processing loops inside the BlockBucket
+    /// via continue_inside() would cause double borrow or require cloning.
+    pub unrolled_block_owner: HashMap<BlockBucketId, Rc<BlockBucket>>,
+    /// Unrolled loop IDs organized by the context where they should be used.
+    pub replacement_context: super::call_unroll_tree::NodeRef,
+}
+
+impl LoopUnrollObserverResult {
+    /// Lookup BlockBucket for the given `key` in the given `cache` or else
+    /// call the `builder` function to create a new BlockBucket and store
+    /// it into `this` and the `cache` and return reference to it.
+    pub(crate) fn get_or_create_unrolled<K, F: FnOnce(&K) -> BlockBucket>(
+        this: &RefCell<LoopUnrollObserverResult>,
+        cache: &RefCell<impl MapLike<K, BlockBucketId>>,
+        key: K,
+        builder: F,
+    ) -> Rc<BlockBucket>
+    where
+        K: std::fmt::Debug,
+    {
+        let id = {
+            // NOTE: This borrow is inside brackets to prevent runtime double borrow error.
```

Comment:
Why does this occur?

Reply:
It doesn't anymore. I accidentally took 2 measures to reduce the scope of this borrow to avoid a BorrowMutError later in the function. Only 1 is needed.